nomad/job_endpoint_statuses_test.go

```diff
@@ -0,0 +1,382 @@
+		name       string
+		qo         structs.QueryOptions
+		jobs       []structs.NamespacedID
+		expect     *structs.Job
+		expectNext uint64 // NextToken (ModifyIndex)
+	}{
+		{
+			name: "page 1",
+			qo: structs.QueryOptions{
+				PerPage: 1,
+			},
+			expect:     jobs[4],
+			expectNext: jobs[3].ModifyIndex,
+		},
+		{
+			name: "page 2",
+			qo: structs.QueryOptions{
+				PerPage:   1,
+				NextToken: strconv.FormatUint(jobs[3].ModifyIndex, 10),
+			},
+			expect:     jobs[3],
+			expectNext: jobs[2].ModifyIndex,
+		},
+		{
+			name: "reverse",
+			qo: structs.QueryOptions{
+				PerPage: 1,
+				Reverse: true,
+			},
+			expect:     jobs[0],
+			expectNext: jobs[1].ModifyIndex,
+		},
+		{
+			name: "filter",
+			qo: structs.QueryOptions{
+				Filter: "ID == " + jobs[0].ID,
+			},
+			expect: jobs[0],
+		},
+		{
+			name: "specific",
+			jobs: []structs.NamespacedID{
+				jobs[0].NamespacedID(),
+			},
+			expect: jobs[0],
+		},
+		{
+			name: "missing",
+			jobs: []structs.NamespacedID{
+				{
+					ID:        "do-not-exist",
+					Namespace: "anywhere",
+				},
+			},
+			expect: nil,
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			resp = doRequest(t, &structs.JobStatusesRequest{
+				QueryOptions: tc.qo,
+				Jobs:         tc.jobs,
+			})
+			if tc.expect == nil {
+				must.Len(t, 0, resp.Jobs, must.Sprint("expect no jobs"))
+			} else {
+				must.Len(t, 1, resp.Jobs, must.Sprint("expect only one job"))
+				must.Eq(t, tc.expect.ID, resp.Jobs[0].ID)
+			}
+			expectToken := ""
+			if tc.expectNext > 0 {
+				expectToken = strconv.FormatUint(tc.expectNext, 10)
+			}
+			must.Eq(t, expectToken, resp.NextToken)
+		})
+	}
+
+	// test blocking queries
+
+	// this endpoint should only unblock if something relevant changes.
+	// "something relevant" is why this seemingly redundant blocking-query
+	// testing is done here, as the logic to determine what is "relevant" is
+	// specific to this endpoint, meaning the latest ModifyIndex on each
+	// job/alloc/deployment seen while iterating, i.e. those "on-page".
+
+	// blocking query helpers
+	startQuery := func(t *testing.T, req *structs.JobStatusesRequest) context.Context {
+		t.Helper()
+		if req == nil {
+			req = &structs.JobStatusesRequest{}
+		}
+		// context to signal when the query unblocks
+		// mustBlock and mustUnblock below work by checking ctx.Done()
+		ctx, cancel := context.WithCancel(context.Background())
+		// default latest index to induce blocking
+		if req.QueryOptions.MinQueryIndex == 0 {
+			idx, err := s.State().LatestIndex()
+			must.NoError(t, err)
+			req.QueryOptions.MinQueryIndex = idx
+		}
+		// start the query
+		// note: queries that are expected to remain blocked leak this goroutine
+		// unless some other test coincidentally frees it up
+		go func() {
+			resp = doRequest(t, req)
+			cancel()
+		}()
+		// give it a moment for the rpc to actually start up
+		// FLAKE ALERT: if this job is flaky, this might be why.
+		time.Sleep(time.Millisecond * 10)
```

Comment:
This is a little unclear. What are we waiting for here? The goroutine to start up? The request to get sent? I'd probably try to gate this with `sync.WaitGroup` instead, or have the function return a channel that the goroutine writes the response onto.

Reply:
the main test that flakes without this is "delete job" (and "delete job on page"), which is supposed to unblock when a job is deleted.  I am fairly sure there is a race between these two things:

* start the blocking query
* delete the job

if the job gets deleted before the query starts blocking, then nothing triggers it to _unblock_, so the test fails.  so what this is waiting for is "everything leading up to blocking on the `WatchSet`" in `rpcHandler.blockingRPC()` (the very last thing in the RPC).  I'm not sure how to slide something in there to signal back to me in the test that it's ready, and 10ms seemed an acceptable sacrifice. :shrug: 

I suppose it doesn't crop up in other `mustUnblock` tests (like "new job") because those operations are slow enough to consistently give the RPC time to start blocking, but I just apply this lil nap across the board.

I could instead sleep down in `t.Run()` between `startQuery()` and `tc.run()` (which deletes the job), but I put it here so it would be right next to the goroutine that spurred me to add it.

phew!  if you have a suggestion for how to boil that down into a tidier comment, I'm happy to hear it! :joy:

Reply:
I added just a little extra here https://github.com/hashicorp/nomad/pull/20130/commits/1b4ce0a5e289fc8c40ffc2c7512e20fcf7556c6c#diff-e1afa22d48b0304823b263b1a4831ec09d3d39b5e9102d2b41535fed4cb5fa5eR240

> // give it a moment for the rpc to actually start up _and begin blocking_

Reply:
Ah, I see what you're doing, and that makes sense. What we've typically done for this in other tests is to put the mutation behind `time.AfterFunc` and then making the blocking query. Ex. in `namespace_endpoint_test.go` you'll see something like:


```go
	// Namespace delete triggers watches
	time.AfterFunc(100*time.Millisecond, func() {
		must.NoError(t, state.DeleteNamespaces(300, []string{ns.Name}))
	})

	req.QueryOptions.MinQueryIndex = 250
	var resp structs.SingleNamespaceResponse
	must.NoError(t, msgpackrpc.CallWithCodec(codec, "Namespace.GetNamespace", req, &resp))
```

That handles both ensuring the index is what you expect and that the delay is long enough. What you've got here is fine, although I'd consider bumping the time up quite a bit because CI machines can have a _lot_ of steal and these take longer than you'd expect.